### PR TITLE
[OWASP] enhance encoding (#3520)

### PR DESCRIPTION
**Pull request by cnsgithub**
Step 1: Copied encoders from owasp-java-encoder

Some notes:
- decided to leave the package structure of owasp-java-encoder untouched, so we can better differentiate between primefaces and third-party sources; also this should make it easier to maintain (btw: should we also move `BoundedInputStream `from commons.io to its original package `org.apache.commons.io`?) -> added `package-info.java` to document owasp-java-encoder's version
- added the newly added classes to `checkstyle-suppressions.xml` to keep them comparable to https://github.com/OWASP/owasp-java-encoder
- added the newly added classes to license-maven-plugin's exclusion list to keep the original BSD 3.0 license headers (required as per the licensing model)

Next step will be the introduction of `EscapeUtils `proposed by @melloware in https://github.com/primefaces/primefaces/issues/3520#issuecomment-375633571 that just delegates calls to `org.owasp.encoder.Encode`.

### CODE REVIEW

**Comment by cnsgithub** [comment]
There are no more calls to `ComponentUtils.escapeText/escapeEcmaScriptText/escapeXml`. They are all `@Deprecated`. Normally, we could remove them also.

I did some rudimentary testing on the showcase. Primarily the ajax part was of interest for me since it was necessary to adjust `PrimePartialResponseWriterTest` accordingly, since it was a bit lenient in XML escaping. At first glance everything is working as expected. Developer console did not show any errors as well.

What I was not sure about is `XmlUtils.escapeTag `which does some special escaping by replacing illegal values with specific separators. To not break compatibility with data export consumers, I didn't touch it yet.

**Comment by cnsgithub** [comment]
`p:button href` mentioned in https://github.com/primefaces/primefaces/issues/3520#issue-307963031 is no longer vulnerable to XSS.

**Comment by cnsgithub** [comment]
Reviews are very welcome.

File: src/main/java/org/owasp/encoder/ASCIIBits.java
Comment (melloware) [inline src/main/java/org/owasp/encoder/ASCIIBits.java]: @tandraschko I am wondering if we should rename the package to org.primefaces.encoder only to not conflict Classloader if someone does include the OWASP JAR itself.  The copyright and all text stays so people will know where this code came from, I just worry about Classloader clollisions. Your thoughts?

File: checkstyle-suppressions.xml
Comment (melloware) [inline checkstyle-suppressions.xml]: Change to org.primefaces.encoder?

**Comment by tandraschko** [comment]
Could someone check the legal approach of copying the class into PF? With Apache2, it's no problem...

**Comment by tandraschko** [comment]
It might hard to say: PF is Apache2 licensed but contains several classes form other sources with other licenses...

**Comment by melloware** [comment]
OWASP is BSD3 licensed which is actually even more friendly than Apache 2.0.  We can basically do whatever we want with the code and it only implies these 3 restrictions:


    * Redistributions of source code must retain the above
      copyright notice, this list of conditions and the following
      disclaimer.
PASSED - we included it

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials
      provided with the distribution.
I THINK WE JUST NEED TO INCLUDE IN OUR LICENSE WE ALSO INCLUDE BSD3.0 licensed software

    * Neither the name of the OWASP nor the names of its
      contributors may be used to endorse or promote products
      derived from this software without specific prior written
I THINK WE WILL PASS THIS ONCE WE REMOVE THE WORD "OWASP" FROM THE PACKAGE NAME.
      permission.

**Comment by cnsgithub** [comment]
> I THINK WE WILL PASS THIS ONCE WE REMOVE THE WORD "OWASP" FROM THE PACKAGE NAME.

In my understanding we're fully conform to this clause because we do not promote PrimeFaces by saying 'powered by OWASP' or anything like that. 

Apache's legal FAQ (http://www.apache.org/legal/resolved.html#category-a) states:
> FOR THE PURPOSES OF BEING INCLUDED IN AN APACHE PRODUCT, WHICH LICENSES ARE CONSIDERED TO BE SIMILAR IN TERMS TO THE APACHE LICENSE 2.0?
> Works under the following licenses may be included within Apache products:
> BSD 3-clause

File: src/main/java/org/owasp/encoder/ASCIIBits.java
Comment (cnsgithub) [reply src/main/java/org/owasp/encoder/ASCIIBits.java]: Good point. But that may also happen with other dependencies not included in the primefaces JAR.

Just some thoughts: 
- Can JAR sealing help us here?
- Can we leverage a tool like [Apache Maven Shade](http://maven.apache.org/plugins/maven-shade-plugin/)?

Comment (melloware) [reply src/main/java/org/owasp/encoder/ASCIIBits.java]: I thought Maven Shade plugin just basically created an uberjar from your dependencies but it still just includes them normally so its no different from a classloader perspective if two jars both have org.owasp.encoder the classloader has to pick one.   I am not familiar with JAR sealing?

Comment (cnsgithub) [reply src/main/java/org/owasp/encoder/ASCIIBits.java]: I just rethought: JAR sealing would not help us here.

However, I'll commit to this PR branch an integration of the shade plugin, be prepared for the upcoming commit comment.

**Comment by cnsgithub** [comment]
Regarding the maven shade plugin:
I already have a running version, albeit not yet commited due to a problem with missing license in owasp-java-encoder. I just filed an issue: https://github.com/OWASP/owasp-java-encoder/issues/12

This is how it basically works:
- no manual copying of owasp-java-encoder sources to PrimeFaces source needed
- instead add a regular compile time dependency to it in the pom (better maintainability)
- use maven-shade-plugin to include owasp-java-encoder classes as well as the license (currently missing) and relocate `org.owasp.encoder package `to shaded.`org.owasp.encoder` to avoid classloading problems like so:
```xml
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-shade-plugin</artifactId>
                <version>3.1.0</version>
                <executions>
                    <execution>
                        <phase>package</phase>
                        <goals>
                            <goal>shade</goal>
                        </goals>
                        <configuration>
                            <transformers>
                                <transformer implementation="org.apache.maven.plugins.shade.resource.IncludeResourceTransformer">
                                    <resource>META-INF/LICENSE_OWASP_ENCODER</resource>
                                    <file>META-INF/LICENSE</file>
                                </transformer>
                            </transformers>
                            <relocations>
                                <relocation>
                                    <pattern>org.owasp.encoder</pattern>
                                    <shadedPattern>shaded.org.owasp.encoder</shadedPattern>
                                </relocation>
                            </relocations>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```
- the JAR root will look like
image

**Comment by melloware** [comment]
That is slick @cnsgithub That way we can just include the dependency and it will bring the code in and also protect the Classloaded since these will be called shaded.

**Comment by cnsgithub** [comment]
Last commit introduced `maven-shade-plugin` as well as `license-maven-plugin` to include and relocate `owasp-java-encoder` files and to include third-party-licenses in primefaces JAR (also see https://github.com/OWASP/owasp-java-encoder/issues/12#issuecomment-378527696).

The JAR structure will look like this now:
```
|   THIRD-PARTY.txt
|   
+---META-INF
|   |   
|   \---resources
|       |   licenses.xml
|       |   
|       +---licenses
|       |       apache license v2 - license-2.0.html
|       |       bouncy castle licence - licence.html
|       |       bsd license - bsd-3-clause.txt.html
|       |       cddl + gplv2 with classpath exception - cddl+gpl.html
|       |       common development and distribution license (cddl) plus gpl - cddl+gpl.html
|       |       common development and distribution license (cddl) v1.0 - cddlv1.0.html
|       |       common development and distribution license (cddl) version 1.0 - cddl.html
|       |       mit license - mit-license.php.html
|       |       mozilla public license - mpl-1.1.html
|       |       similar to apache license but with the acknowledgment clause removed - license.txt
|       |       the apache software license, version 2.0 - license-2.0.txt
|       |       the bsd 3-clause license - bsd-3-clause.txt.html
|       |       
\---primefaces
    \---shaded
        \---org
            \---owasp
                \---encoder
                        ASCIIBits.class
                        CDATAEncoder.class
                        ChainedEncoder.class
                        CSSEncoder$Mode.class
                        CSSEncoder.class
                        Encode$Buffer.class
                        Encode.class
                        EncodedWriter.class
                        Encoder.class
                        Encoders.class
                        HTMLEncoder.class
                        JavaEncoder.class
                        JavaScriptEncoder$Mode.class
                        JavaScriptEncoder.class
                        Unicode.class
                        UnsupportedContextException.class
                        URIEncoder$Mode.class
                        URIEncoder.class
                        XMLCommentEncoder.class
                        XMLEncoder$Mode.class
                        XMLEncoder.class
```

E.g. `THIRD-PARTY.txt` contains this line:
`     (The BSD 3-Clause License) Java Encoder (org.owasp.encoder:encoder:1.2.1 - https://www.owasp.org/index.php/OWASP_Java_Encoder_Project/encoder)`

The original license texts are copied to `META-INF/resources/licenses`, `licenses.xml` contains a summary.

@tandraschko @melloware What do you think? Should we give this PR a try?

**Comment by tandraschko** [comment]
Thats not only up to me, i need to discuss this before.
And currently i don't have much time, i will priorise the older PRs before.

**Comment by cnsgithub** [comment]
@tandraschko It's two months later now... Are there any plans to merge it?

**Comment by tandraschko** [comment]
No Time currenty, sry. Need also discuss thus First.

**Comment by tandraschko** [comment]
@cnsgithub really would like to discuss this with the PrimeTeam, but even all of my private discussion-topics are not all done till today.
I promise i will take care of it before 6.3 but WDYT about doing it in 3 steps?
1) Introduce EscapeUtils, move all of our existing methods there
2) Correct all usages to use the right method there
3) switch to OWASP in the EscapeUtils

It would be much better to follow all the changes.

**Comment by cnsgithub** [comment]
@tandraschko Thanks for the feedback. I think the problem of point 2 is, that there isn't an adequate encoding implementation for each context right now. So it makes no big difference if we just stay with the old method names or refactor existing code to use newly introduced ones. All calls to the OWASP encoders are already wrapped in `EscapeUtils`. I'm not interested in repeatedly refactoring code that has been proposed for three months now (it has conflicts again now). I also do not have enough time to do so. I think PrimeTek should be interested in merging things like that in own initiative.

**Comment by tandraschko** [comment]
Triggered the discussion with PrimeTek now.

Just 2 things before i could merge after the discussion:
1) please don't change our imports, it's probably your IDE setting about grouping imports
2) fix the conflicts

**Comment by cnsgithub** [comment]
Resolved the conflicts.

File: pom.xml
Comment (Rapster) [inline pom.xml]: Using `compile` does not make that dependency transitive? Or shade will remove this dependency from the pom.ml

**Comment by tandraschko** [comment]
Conflicts now - can you resolve them? (I'm still waiting for feedback from PrimeTek)

**Comment by tandraschko** [comment]
Got the GO ;)
If you can fix the conflicts, i can merge it early next before holiday.

About the transitive dependency, it would should be fine if we set "<optional>true</optional>"

**Comment by Rapster** [comment]
What is the policy about dependencies? IMO, it makes sense to use shade if we use a few classes over plenty. In this case, we use all classes (according to https://github.com/primefaces/primefaces/pull/3525#issuecomment-378541688), why not using it as a true dependency then? In general, I would use shading whenever copying a class from another repo is needed.

@cnsgithub you asked
> (btw: should we also move `BoundedInputStream` from commons.io to its original package org.apache.commons.io?)

Yes it makes sense, but if we use last version of commons-io (which is 2.5) it will copy a lot more than just `BoundedInputStream` because it has a "dependency" to `IOUtils` (see https://github.com/apache/commons-io/blob/master/src/main/java/org/apache/commons/io/input/BoundedInputStream.java) It'll copy a bunch of classes just because `BoundedInputStream` needs `IOUtils#EOF`. Although, it should be fine using `commons-io` 2.4

When using `maven-shade-plugin` we just have to make sure, shade dependencies are no longer transitive (removed from the final pom.xml)

Another thing, Thomas said

> please don't change our imports, it's probably your IDE setting about grouping imports

So far, imports are sorted in this order:

- java packages
- javax packages
- others (like org.primefaces)

**Comment by tandraschko** [comment]
Cagatay only allowed me to shade it - because it's used by many core features and PF goal is 1jar = everything cool :D

-> we can use OWASP if we shade it, everything else needs to be untouched (like the BoundedInputStream or json).

**Comment by melloware** [comment]
Personally I prefer transitive dependencies over shaded ones and here is why.  If for some security reason a user wants to override the version of OWASP or (JSON future) they can simply doing by updating their pom.xml and excluding the versions included with PF.  If however you wrap it up inside the PF JAR there is no way to override anymore without ripping the jar open and removing classes.

**Comment by Rapster** [comment]
@melloware Yes, but it has also dowsides:

- for PF, what if the upgrading version of the dependency (overrided by the end user) is not compatible with PF (projects using guava suffered a lot about this), 

- or the project itself , what if PF is actually overriding the dependency of the project, this an create regressions.

**Comment by melloware** [comment]
Agreed it definitely has its pro's and cons.

**Comment by melloware** [comment]
OK that settles it we shade it :)

**Comment by Rapster** [comment]
@cnsgithub you should use `minimizeJar` option, it will copy only what you need :wink:

**Comment by Rapster** [comment]
> -> we can use OWASP if we shade it, everything else needs to be untouched (like the BoundedInputStream or json).

Why the approach should be different?

**Review by Rapster — CHANGES_REQUESTED** [review]
What is been shaded does not need to be included in checkstyle or license plugin, or it does?

**Comment by tandraschko** [comment]
AFAIK checkstyle checks the source and shading uses the compiled classes? so nn to exclude something actually?!

**Comment by Rapster** [comment]
Yes no need, as an example, I had a try shading json (which does not follow our checkstyle) and had no issues with checkstyle plugin after excluding

**Comment by cnsgithub** [comment]
@Rapster I removed the checkstyle and license exclusions as you suggested.

This PR was quite a time ago and it wasn't that easy to merge, so hopefully I did not miss something. Sorry, not much time currently...

**Comment by tandraschko** [comment]
Looks like you readded getHrefURL getResourceURL isValueBlank to ComponentUtils - please check this changes

**Comment by cnsgithub** [comment]
Removed re-added methods.

**Comment by cnsgithub** [comment]
There are some failing tests in `NativeUploadedFileTest`. Those are however not related to this PR.

**Comment by Rapster** [comment]
@mertsincan Can you have a look please? The problem is still there

**Comment by Rapster** [comment]
I wish few of my feedbacks would be answered before merging...

**Comment by tandraschko** [comment]
the codechanges are fine, just pom needs a final tuning now because of the license
i removed as the it broke the build on my machine

**Comment by tandraschko** [comment]
@cnsgithub thanks for the changes!
Can you also migrate all code to EscapeUtils?

**Comment by cnsgithub** [comment]
That should have been done already, except for new code, if any. We could see that very easily by just dropping the deprecated escape methods in `ComponentUtils`.

**Comment by Rapster** [comment]
I left a feedback regarding shaded naming policy, because it's gonna happen again with other library. I thought of `org.primefaces.owasp.encoder` just like we have `org.primefaces.json`, but instead it's `primefaces.shaded.org.owasp.encoder` @melloware left a comment I didn't understand, so I'd like to understand it to make sure I don't miss anything.

**Comment by cnsgithub** [comment]
I don't like the idea of mixing up the namespaces. However, feel free to use a prefix of your choice.

**Comment by Rapster** [comment]
It's not a matter of personal choice or taste, the idea is to follow guidelines of what PrimeFaces have done so far. I consider json as a shaded dependency (even though it's not entirely true), so I would proceed in the very same way with owasp (this is why I proposed `org.primefaces.owasp.encoder`) . Unless we think `primefaces.shaded` is the good prefix to go with, I'm totally up for it as long we keep doing that with the next dependencies (e.g org.json) I just don't want to have some shaded stuff on one side and the other shaded stuff on the other side... 

But Thomas seems to not mind it, so I guess there is no need to continue this conversation. Next time, please, answer the feedback itself :wink:

**Comment by cnsgithub** [comment]
I generally also prefer consensus over personal taste. But it has been a long time, almost half a year. And there was not so much feedback from you regarding this point at this time. And in fact I still have problems to find your review containing this shading namespace proposal. I think sometimes we should rather concentrate on the issues than discussing trivial things that can be changed without efforts in the future anyway. owasp was the first lib to be shaded, other libs may follow these rules or redefine them, not the other way around.

**Comment by tandraschko** [comment]
i changed it to org.primefaces.shaded.owasp.encoder for now

we should use "org.primefaces." prefix everywere i think it makes sense to use a "shaded" package
we can move json for 7.0, too

**Comment by melloware** [comment]
@Rapster my comment has been answered I think from reading all the changes I am good to go with all the fixes here!

**Comment by Rapster** [comment]
@melloware I didn't find it, but I can figure out what you and cnsgithub meant :wink:

@tandraschko Cool! I'll create an issue so we don't forget about this for 7.0

**Comment by melloware** [comment]
The only thing I see now in Maven after updating is this...

```
[INFO] --- license-maven-plugin:1.16:aggregate-download-licenses (aggregate-download-licenses) @ primefaces ---
[WARNING] Unable to retrieve license for dependency: org.owasp.esapi:esapi
[WARNING] http://creativecommons.org/licenses/by-sa/3.0/
[WARNING] Server returned HTTP response code: 403 for URL: http://creativecommons.org/licenses/by-sa/3.0/
```

**Pull request merged**
